plugin/evm/vm.go

```diff
@@ -738,22 +660,92 @@ func (vm *VM) awaitTxPoolStabilized() {
 	}
 }
 
+// needToBuild returns true if there are outstanding transactions to be issued
+// into a block.
+func (vm *VM) needToBuild() bool {
+	size, err := vm.chain.PendingSize()
+	if err != nil {
+		log.Error("Failed to get chain pending size", "error", err)
+		return false
+	}
+	return size > 0 || vm.mempool.Len() > 0
+}
+
+// buildEarly returns true if there are sufficient outstanding transactions to
+// be issued into a block to build a block early.
+func (vm *VM) buildEarly() bool {
+	size, err := vm.chain.PendingSize()
+	if err != nil {
+		log.Error("Failed to get chain pending size", "error", err)
+		return false
+	}
+	return size > batchSize || vm.mempool.Len() > 1
+}
+
+// buildBlockTwoStageTimer is a two stage timer that sends a notification
+// to the engine when the VM is ready to build a block.
+// If it should be called back again, it returns the timeout duration at
+// which it should be called again.
+func (vm *VM) buildBlockTwoStageTimer() (time.Duration, bool) {
+	vm.buildBlockLock.Lock()
+	defer vm.buildBlockLock.Unlock()
+
+	switch vm.buildStatus {
+	case dontBuild:
+		return 0, false
+	case conditionalBuild:
+		if !vm.buildEarly() {
+			vm.buildStatus = mayBuild
+			return (maxBlockTime - minBlockTime), true
+		}
+	case mayBuild:
+	case building:
+		// If the status has already been set to building, there is no need
+		// to send an additional request to the consensus engine until the call
+		// to BuildBlock resets the block status.
+		return 0, false
+	}
+
+	select {
+	case vm.notifyBuildBlockChan <- commonEng.PendingTxs:
+	default:
+		log.Error("Failed to push PendingTxs notification to the consensus engine.")
+	}
+	vm.buildStatus = building
```

Comment:
Should we only mark as building if we could push to `notifyBuildBlockChan`?

Reply:
My thinking on this was that since this is the only message type, it should be in building already if the channel is full, but this is probably not a future proof assumption. Will update.